examples/job_openapi.rs

```diff
@@ -16,33 +22,40 @@ async fn main() -> anyhow::Result<()> {
     let client = APIClient::new(config);
 
     // Create a Job
-    let my_job = json!({
-        "apiVersion": "batch/v1",
-        "kind": "Job",
-        "metadata": {
-            "name": "empty-job"
+    let my_job = Object {
+        types: TypeMeta {
+            apiVersion: Some("batch/v1".to_string()),
+            kind: Some("Job".to_string()),
+        },
+        metadata: ObjectMeta {
+            name: "empty-job".to_string(),
+            ..Default::default()
         },
-        "spec": {
-            "template": {
-                "metadata": {
-                    "name": "empty-job-pod"
-                },
-                "spec": {
-                    "containers": [{
-                        "name": "empty",
-                        "image": "alpine:latest"
+        spec: JobSpec {
+            template: PodTemplateSpec {
+                metadata: Some(OpenApiObjectMeta {
```

Comment:
This seems like a pretty big API wart to expose.. :/

Reply:
Yeah, the json! macro is pretty nice to avoid having to explicitly type out all the defaults.

We could maybe do some magic with TryFrom json::Value -> relevant K, but that would only give us runtime type-safety, which only saves us from having to actually post it to the k8s api.

Reply:
I was mostly talking about the `ObjectMeta` vs `OpenApiObjectMeta` distinction.

The `Default` proliferation sounds like it could be worked around with a macro that implicitly adds a `..Default::default()` member to every object declaration? Alternatively, builders..

Reply:
Possibly. Just keep in mind that we need to let people create incomplete objects anyway for server-side-apply (we set the fields we care about, api server handles the rest). I think it would be a lot easier if they could just use the `json!` macro (or something similar for easy yaml if it exists) for it.

Reply:
Yeah.. I considered partial updates (apply, patch) out of scope for this PR.

> I think it would be a lot easier if they could just use the json! macro (or something similar for easy yaml if it exists) for it.

Not a huge fan of telling users to send untyped blobs.. I guess we could `impl<K> SerializeKubeObject<K> for serde_json::Value`, so they at least get types that last that far...

Reply:
Hm, for server-side apply you'd usually want to always send the same fields (the fields that you want to manage), so in that case you should probably define a `FooDeploymentSpec` (or whatever other resource) struct that contains the subset of fields that you want to manage.

Reply:
That might work for the Spec, but not sure it's that clear cut most of the time though. I mean, most of the time you often just end up setting all these peripheral properties like labels, annotations, finalizers, ownerReferences anyway. And for an actual CRD you manage in a controller, you often just read the Spec, and patch the Status.

If you have multiple controllers annotating the .status object of the same root resource, then they need to not step on each others toes. I can imagine that's usually fine if they define mutually exclusive subset structs, but can also imagine scenarios where you may want to read some properties in both of them, but still have one controller be the authoritative source of every field.

Patching/server-side-apply feels like an acceptable / least privilege way of doing that, without having to implement something similar to client-go's `RetryOnConflict`. It's possible we can still enforce some type safety on it by using the [fieldManager param](https://kubernetes.io/docs/reference/using-api/api-concepts/#managers) though.

Reply:
FWIW I tried implementing an `auto_default!` macro to reduce some of the boilerplate: https://github.com/Appva/kube-rs/commit/3c5a9a1649d1e5418b800a2c5fd9264fe196417d

Reply:
Late response. Sorry.
Yeah, the owned subset thing might be possible (although you can't have data members in traits atm afaik). But I'm still not convinced we should be serialising all our state everytime. The Go world is doing a lot of individual updating to huge objects to avoid stepping on the toes of other controllers, but also to avoid update conflicts. Replace methods won't even let you do things without an existing version, so they are kind of useless for most practical applications (because you effectively have to do a .get(), merge your data onto that, then do a .replace()).

So I am a bit scared to deviate.

But it also feels very impractical to me.
Like `Deployment.spec.template.metadata.labels`. If you needed to update a field that deep you need to define the whole object tree. If you wrote it out with `json!` the existing kubernetes admission controllers would catch bad writes (albeit at runtime, which you can test - but more costly).

Reply:
I should probably have clarified.. `OwnedSubset` itself doesn't care about the data itself, it's just a glorified `From<K>`. The macro would generate the (equivalent of) `into`, enforcing that all fields are `OwnedSubset`s of their corresponding fields in the `base`. That should work as long as we also blanket `impl<K> OwnedSubset<K> for K` (for the leaves).

Reply:
Then again, [k8s-openapi seems to go out of their way to avoid serializing empty fields](https://docs.rs/k8s-openapi/0.7.1/src/k8s_openapi/v1_17/api/apps/v1/deployment_spec.rs.html#139), so all of this extra machinery probably wouldn't actually help. From what I can tell, just `Api::apply(serde_json::to_vec(&deployment))` should actually do the right thing? Then again, I'm probably missing something and would have to experiment more to get anywhere.

Reply:
I should also clarify my own needs before kicking off a huge over-engineering project: at the moment I mostly care about controllers that fully own their resources, `create`, `replace`, and `watch` are what I really care about.

Reply:
> Then again, [k8s-openapi seems to go out of their way to avoid serializing empty fields](https://docs.rs/k8s-openapi/0.7.1/src/k8s_openapi/v1_17/api/apps/v1/deployment_spec.rs.html#139), so all of this extra machinery probably wouldn't actually help. From what I can tell, just `Api::apply(serde_json::to_vec(&deployment))` should actually do the right thing? Then again, I'm probably missing something and would have to experiment more to get anywhere.

Oh wow. `k8s-openapi` really have gotten a lot smarter.

Yeah, if with had `Api::apply` as an alias for `patch` with server-side apply, then that should do the right thing. Particularly since we get to take advantage over that serde impl in k8s-openapi.

Reply:
> I should also clarify my own needs before kicking off a huge over-engineering project: at the moment I mostly care about controllers that fully own their resources, `create`, `replace`, and `watch` are what I really care about.

If you have any defaulting behaviour from the openapi schema in the CRD, or any mutating admission controllers, then using server-side `Api::apply` instead of both `replace` and `create` would probably be the best thing (as that preserves the behaviour of those). But in your simpler case, you __can__ just use `Api::create` and `Api::replace` in the simple case.